### ISSUE TITLE
Ruby cannot be executed from a shoes script

### ISSUE THREAD

**Issue by lmrodriguezr** [shoes/shoes4#1184]
I'm starting to feel like I'm contributing nothing but edge cases. Still, here is the problem (?) I found.

`Shoes.app` seems to allow for system calls without problems (except in .app packages for OS X, but I'll open a new thread for that once I understand the problem better). Everything is allowed, it seems, EXCEPT ruby! Here, take a look at this little app:

``` ruby
Shoes.app do
   stack margin: 40 do
      title "Shell testing"
      flow do
         commands = ["echo 123", "perl -e 'print 234'", "ruby -e 'puts 345'",
            "/usr/bin/ruby -e 'puts 567'", "jruby -e 'puts 456'",
            "which ruby", "which jruby"]
         commands.each do |c|
            button c do
               t0 = Time.now
               $e.text = `#{c} 2>&1`
               t1 = Time.now
               $t.text = "It took #{(t1-t0)*1000}ms."
            end
         end
      end
      $e = edit_box(width:1.0, height:250)
      $t = para "", weight:"heavy", align:"right"
   end
end
```

Both `echo` and `perl` work fine, and I've tested redirections, pipes, and other stuff and everything seems to be fair game. However, if I call ruby (whether I just use `ruby` or I use the full path, which in my case is `/usr/bin/ruby`) I get [this perverse message](http://pastebin.com/E018cQn8) complaining about `jruby` not having the `pry` gem. At this point, I should probably mention that in my system `ruby` is Ruby, not JRuby:

``` bash
$ ruby -v # same as /usr/bin/ruby -v
ruby 2.0.0p481 (2014-05-08 revision 45883) [universal.x86_64-darwin14]
```

Even more strange, though, `jruby` works! The beef seems to be only with `ruby`, which gets strangely replaced by `jruby` even if I call the full path. Something else I noticed is that the `jruby` call takes a very long time (not a big problem, but perhaps an extra clue). Compare the **~2ms** of `echo` or `perl` and the **~20ms** of `ruby` and `/usr/bin/ruby` (before failing) with the **~7sec** of `jruby`. The same jruby command takes **~2.5sec** in the terminal.

**Comment by paynito** [comment]
I'm not an expert but maybe you ran `rvm use jruby; ruby -v `   and that makes bash/zsh call jruby all the time instead or ruby?

**Comment by PragTob** [comment]
Hey there,

it is great that you contribute all this edge cases and make such detailed reports about them. That is a great help for us and please continue to do so - it is really appreciated :heart: 

About the problem, it is very weird. That might as well be an upstream problem (with JRuby) or a problem very specific to your system setup. We'll try to figure that out.

JRuby taking so long to start is sadly rather normal (although the difference is quite interesting).

Which ruby version manager do you use? rvm? rbenv? chruby? You're on OSX, right?

Cheers,
Tobi

**Comment by PragTob** [comment]
@lmrodriguezr can't reproduce over here, works like a charm. Couple of ideas I have:

* might be related to the OSX -XStartOnFirstThread thing ( @jasonrclark might you be so kind and check? :) )
* `/usr/bin/ruby` is likely just a symlink to the real binary, which ruby version manager are you using?
* what is the output of `ruby -v` right before you launch the shoes app?

I have no ideas why it tries to find pry and use bundler. My only idea is that you might run it from within a project directory where you have a `Gemfile.lock` and that has the pry gem in it and some magic tries to pick that up.

Tobi

**Comment by jasonrclark** [comment]
Tried to duplicate things on my Mac, but didn't get the same problems either :(

So I did observe that `ruby` properly found my current Ruby via my version manager (`rbenv`) while `/usr/bin/ruby` was my system-installed Ruby.

These lines in particular in that error look weird... why do we go from the system Ruby over into a homebrew installed JRuby???

```
...
        from /usr/local/Cellar/jruby/9.0.3.0/libexec/lib/ruby/gems/shared/gems/bundler-1.10.6/lib/bundler/setup.rb:18:in `<top (required)>'
        from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:55:in `require'
        from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:55:in `require'
```

**Comment by PragTob** [comment]
@lmrodriguezr could you please try this out in a pure ruby script and tun that with JRuby to see if the problem persists?

**Comment by lmrodriguezr** [comment]
> @PragTob:
> My only idea is that you might run it from within a project directory where you have a Gemfile.lock

I think that was it! Almost... I actually removed `Gemfile` and `app.yaml`and it worked fine. So I restated [`Gemfile`](https://github.com/shoes/shoes4/blob/master/Gemfile), and voilà, the error is back! The thing is... `Gemfile` is necessary for proper packaging, but its very presence seems to be the culprit of this mishap. So the only workaround I could find was to move `Gemfile` to a temporary location, and restate it only when I need to package apps. BTW, the problem's there whenever I run shoes from within the shoes4 dir and `Gemfile` is present, even if I run the external `shoes` executable from the installed gem (instead of `bin/shoes`).

> @PragTob:
> could you please try this out in a pure ruby script and tun that with JRuby to see if the problem persists?

It works without problems:
```bash
$ jruby -e 'puts `/usr/bin/ruby -e "puts 123" 2>&1`'
123
$ jruby -e 'puts `ruby -e "puts 123" 2>&1`'
123
```

@jasonrclark Yes, that's what's puzzling me: why is it loading homebrew's jruby??? Even when I call it with the full path as `/usr/bin/ruby`!

BTW: Yes, I'm using OS X, and I haven't installed any version managers. I'm using the `ruby` off the OS X distro, and jruby from homebrew.

**Comment by lmrodriguezr** [comment]
@PragTob the jruby test on pure ruby works fine regardless of `Gemfile` or location.

**Comment by PragTob** [comment]
Thank you for trying all that out! I'm unsure how to continue here however... i.e. no idea how to really fix this. @jasonrclark do you have an idea?

**Comment by lmrodriguezr** [comment]
Yeah, I'm not sure how to proceed either. Perhaps the existence of this thread alone is sufficient, so somebody with the same problem in the future might find a workaround (?). Just in case this helps, here's a session with a minimal test that still causes the problem whenever `Gemfile` is present:
```bash
shoes4 $ # My environmental ruby/jruby:
shoes4 $ ruby -v
ruby 2.0.0p481 (2014-05-08 revision 45883) [universal.x86_64-darwin14]
shoes4 $ jruby -v
jruby 9.0.3.0 (2.2.2) 2015-10-21 633c9aa Java HotSpot(TM) 64-Bit Server VM 25.66-b17 on 1.8.0_66-b17 +jit [darwin-x86_64]
shoes4 $ which ruby
/usr/bin/ruby
shoes4 $ which jruby
/usr/local/Cellar/jruby/9.0.3.0/libexec/bin/jruby
shoes4 $ # Test 1: a small program (fails)
shoes4 $ cat minitest.rb
$stderr.puts `/usr/bin/ruby -e 'puts 123' 2>&1`
shoes4 $ bin/shoes-swt minitest.rb
/usr/local/Cellar/jruby/9.0.3.0/libexec/lib/ruby/gems/shared/gems/bundler-1.10.6/lib/bundler/resolver.rb:347:in `block in verify_gemfile_dependencies_are_found!': You have requested: (Bundler::GemNotFound)
  pry >= 0

The bundle currently has pry locked at 0.10.3.
Try running `bundle update pry`
	from /usr/local/Cellar/jruby/9.0.3.0/libexec/lib/ruby/gems/shared/gems/bundler-1.10.6/lib/bundler/resolver.rb:325:in `each'
	from /usr/local/Cellar/jruby/9.0.3.0/libexec/lib/ruby/gems/shared/gems/bundler-1.10.6/lib/bundler/resolver.rb:325:in `verify_gemfile_dependencies_are_found!'
	from /usr/local/Cellar/jruby/9.0.3.0/libexec/lib/ruby/gems/shared/gems/bundler-1.10.6/lib/bundler/resolver.rb:198:in `start'
	from /usr/local/Cellar/jruby/9.0.3.0/libexec/lib/ruby/gems/shared/gems/bundler-1.10.6/lib/bundler/resolver.rb:182:in `resolve'
	from /usr/local/Cellar/jruby/9.0.3.0/libexec/lib/ruby/gems/shared/gems/bundler-1.10.6/lib/bundler/definition.rb:200:in `resolve'
	from /usr/local/Cellar/jruby/9.0.3.0/libexec/lib/ruby/gems/shared/gems/bundler-1.10.6/lib/bundler/definition.rb:140:in `specs'
	from /usr/local/Cellar/jruby/9.0.3.0/libexec/lib/ruby/gems/shared/gems/bundler-1.10.6/lib/bundler/definition.rb:185:in `specs_for'
	from /usr/local/Cellar/jruby/9.0.3.0/libexec/lib/ruby/gems/shared/gems/bundler-1.10.6/lib/bundler/definition.rb:174:in `requested_specs'
	from /usr/local/Cellar/jruby/9.0.3.0/libexec/lib/ruby/gems/shared/gems/bundler-1.10.6/lib/bundler/environment.rb:18:in `requested_specs'
	from /usr/local/Cellar/jruby/9.0.3.0/libexec/lib/ruby/gems/shared/gems/bundler-1.10.6/lib/bundler/runtime.rb:13:in `setup'
	from /usr/local/Cellar/jruby/9.0.3.0/libexec/lib/ruby/gems/shared/gems/bundler-1.10.6/lib/bundler.rb:127:in `setup'
	from /usr/local/Cellar/jruby/9.0.3.0/libexec/lib/ruby/gems/shared/gems/bundler-1.10.6/lib/bundler/setup.rb:18:in `<top (required)>'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/rubygems/core_ext/kernel_require.rb:55:in `require'
shoes4 $ #'
shoes4 $ ruby minitest.rb 
123
shoes4 $ jruby minitest.rb
123
shoes4 $ # Test 2: just the version (works)
shoes4 $ cat minitest2.rb
$stderr.puts `/usr/bin/ruby -v`
shoes4 $ bin/shoes-swt minitest2.rb
ruby 2.0.0p481 (2014-05-08 revision 45883) [universal.x86_64-darwin14]
shoes4 $ # Test 3: a small program, without the Gemfile (works)
shoes4 $ mv Gemfile Gemfile.bu
shoes4 $ bin/shoes-swt minitest.rb
123
```

**Comment by lmrodriguezr** [comment]
[Perhaps the thread name should be changed, since it doesn't really matter if the code is within a `Shoes.app` block, just if it's called by the shoes executable within the shoes folder.]

**Comment by PragTob** [comment]
Thanks for the thorough example, I think this is due to our custome `Bundler.require` code on which @jasonrclark is the expert and I fail to remember why we exactly need it... packaging something?

**Comment by jasonrclark** [comment]
We do the `Bundler.require` to avoid people having to run `bundle exec` on the `shoes` executable--if they are running Shoes in the presence of a `Gemfile`, then we'll pick it up for them automatically.

I'm still super-baffled too, and it'll probably be a bit before I'll investigate further given my schedule and the existence of a workaround. In the meantime a retitling so folks can find it as a workaround is probably in order.

Appreciate all your help in getting diagnostics on this @lmrodriguezr ! :sparkles:

**Comment by jasonrclark** [comment]
Back to reviewing stuff at long last (hurray!) and I think I finally understand the behavior I'm seeing and what you were seeing @lmrodriguezr.

The first key we've already established is that running `ruby` for you is running your system Ruby (CRuby 2.0.0), not `jruby` (which I get because of my version manager). When I run and force to my system ruby by running `/usr/bin/ruby` I get the same Bundler error you do.

So what's the error all about? When the Shoes process backticks to start up Ruby, it passes along its ENV. Part of that ENV causes Ruby to load Bundler (not to self-promote too much, but I gave [a lightning talk about how it accomplishes this which might be of interest](https://www.youtube.com/watch?v=o_VHvwjFcmQ)) and a piece of that is where to find the gems. That is pointed at where Shoes loaded the gems under JRuby, which has a flaming meltdown when trying to be used by the CRuby 2.0.0 in the system.

What can be done about this? I think from a generic Shoes perspective, not much. We've got our environment loaded, if the process you call from there gets an unexpected ENV, we do much about that.

But fortunately, Bundler comes to the rescue with [`Bundler.with_clean_env`](http://bundler.io/v1.3/man/bundle-exec.1.html#Shelling-out). By wrapping this method call around the backticks in you sample, everyone ran as I'd expect them to.

Closing out, and thank you again for the excellent bug reporting. All those details made it possible for me to debug it with the knowledge of Bundler that present-day-Jason now has :sparkles:
